### PR TITLE
Marking more core tests as slow

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

It's starting to take annoyingly long to run `pytest tests -m "core and not slow" -n auto`.  So we need to mark more tests as slow.

**Description of the Change:**

Marked all native mcm tests as slow, all multiprocessing tests as slow, and various other tests discovered with `--durations=50`.

**Benefits:**

Easier to run tests locally.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/ops/functions/matrix.py [hunk 1]
```diff
@@ -233,6 +233,8 @@ def _matrix_transform(
     wires = kwargs.get("device_wires", None) or tape.wires
     wire_order = wire_order or wires
 
+    print(tape.circuit)
+
```
Comment (albi3ro) [inline pennylane/ops/functions/matrix.py:237]:
```suggestion
```

**Comment by trbromley** [comment]
IIRC the slow tests are only run on merges into `master`? If so, one drawback here can the potential for more failures to show up post-merge.

**Comment by josh146** [comment]
We could do something similar to what Lightning does with their GPU tests, and catalyst with their wheel builds.

- GPU/wheel/slow tests are not run on every PR commit.
- there is a GitHub label that triggers the GPU/wheel/slow tests
- The github label is applied right before merge, to ensure that GPU/wheel/slow tests are not broken by the PR.

This helps avoid broken `master` branches, and makes it easier to identify which PR breaks these tests.

**Comment by albi3ro** [comment]
> IIRC the slow tests are only run on merges into `master`? If so, one drawback here can the potential for more failures to show up post-merge.

CI does not use this marker.  It just aids in local development.  For example, I locally run `python -m pytest tests -m "core and not slow" -n auto`.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 2]
```diff
@@ -26,6 +26,8 @@
     parse_native_mid_circuit_measurements,
 )
 
+pytestmark = pytest.mark.slow
```
Comment (Mandrenkov) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:29]: # Praise
Nice, TIL about `pytestmark`!

File: tests/gradients/core/test_pulse_gradient.py [hunk 3]
```diff
@@ -919,7 +919,7 @@ def test_constant_ry(self, dev_name, num_split_times, t):
 
         dev = qml.device(dev_name, wires=1)
         # Effective rotation parameter
-        p = params[0] * (delta_t := (T[-1] - T[0]))
+        p = params[0] * (delta_t := T[-1] - T[0])
```
Comment (Mandrenkov) [inline tests/gradients/core/test_pulse_gradient.py:922]: # Question
Was this recommended by Pylint? I would imagine that
```python
p = x := (y - z)
```
is less ambiguous than
```python
p = x := y - z
```
because it is not obvious that this is _not_ equal to
```python
p = (x := y) - z
```

File: tests/qnn/test_keras.py [hunk 4]
```diff
@@ -922,6 +922,8 @@ def circuit(inputs, w1, w2):
     assert info["device_name"] == "default.qubit"
 
 
+@pytest.mark.slow
+@pytest.mark.tf
```
Comment (Mandrenkov) [inline tests/qnn/test_keras.py:926]: # Praise
Good catch! 🎣

File: tests/devices/default_qubit/test_default_qubit.py [hunk 5]
```diff
@@ -27,6 +27,13 @@
 np.random.seed(0)
 
 
+max_workers_list = [
+    None,
+    pytest.param(1, marks=pytest.mark.slow),
+    pytest.param(2, marks=pytest.mark.slow),
```
Comment (Mandrenkov) [inline tests/devices/default_qubit/test_default_qubit.py:33]: # Praise
Nice, I didn't know you could apply markers on a per-parameter basis! 🎉

File: tests/devices/default_qubit/test_default_qubit.py [hunk 6]
```diff
@@ -165,7 +172,7 @@ def test_doesnt_support_other_gradient_methods(self, gradient_method):
 class TestBasicCircuit:
     """Tests a basic circuit with one RX gate and two simple expectation values."""
 
-    @pytest.mark.parametrize("max_workers", [None, 1, 2])
+    @pytest.mark.parametrize("max_workers", max_workers_list)
```
Comment (Mandrenkov) [inline tests/devices/default_qubit/test_default_qubit.py:175]: # Suggestion
Have we considering replacing all of these method-level parameterizations with a single parameterization of the test class?

File: tests/transforms/test_optimization/test_pattern_matching.py [hunk 7]
```diff
@@ -610,6 +610,7 @@ def mod_5_4():
 
         assert np.allclose(qml.matrix(optimized_qnode)(), qml.matrix(qnode)())
 
+    @pytest.mark.slow
```
Comment (Mandrenkov) [inline tests/transforms/test_optimization/test_pattern_matching.py:613]: # Note
This one took over a minute on my machine! 😲

**Review by Mandrenkov — APPROVED** [review]
Looks good to me! I didn't know about `--durations` - very useful stuff!

File: tests/gradients/core/test_pulse_gradient.py (hunk 3 above)
Comment (albi3ro) [reply tests/gradients/core/test_pulse_gradient.py:922]: it was a pylint thing.

**Comment by Mandrenkov** [comment]
The improvement in execution time is very noticeable!

I ran a simple benchmark using:
```shell
$ time python -m pytest tests -m "core and not slow" -n auto
```

The difference is nearly an order of magnitude!
| Branch | Output |
| :--- | :--- |
| `master` | `3987.75s user 379.45s system 269% cpu 26:57.56 total` |
| `more-slow-tests` | `873.11s user 69.65s system 508% cpu 3:05.24 total` |

**Review by Alex-Preciado — APPROVED** [review]
